misk-jobqueue/src/main/kotlin/misk/jobqueue/JobQueue.kt

```diff
@@ -12,17 +13,23 @@ interface JobQueue {
   /**
    * Enqueue a job onto the given queue, along with a set of job attributes.
    *
-   * @param queueName the name of the queue on which to place the job
+   * @param queueName The name of the queue on which to place the job.
    * @param body The body of the job; can be any arbitrary string - it is up to the enqueuer and
-   * consumer to agree on the format of the body
+   * consumer to agree on the format of the body.
+   * @param idempotenceKey Client-assigned unique key, useful for application code to detect duplicate work.
+   * Implementations of both [JobQueue] and [JobConsumer] are expected to _not_ perform any filtering based on this
+   * value, as it carries meaning only for application code (i.e. any logic around this property should take place in
+   * [JobHandler]s). Defaults to a randomly generated UUID when not explicitly set.
    * @param deliveryDelay If specified, the job will only become visible to the consumer after
    * the provided duration. Used for jobs that should delay processing for a period of time.
-   * @param attributes Arbitrary contextual attributes associated with the job
+   * @param attributes Arbitrary contextual attributes associated with the job. Implementations may limit the number of
+   * attributes per message.
    */
   fun enqueue(
     queueName: QueueName,
     body: String,
+    idempotenceKey: String = UUID.randomUUID().toString(),
```

Comment:
this is going to break existing clients because it's not binary compatible. you need to create another overloaded method or can possibly use `@JvmOverloads`

Reply:
I wasn't familiar with `@JvmOverloads`  before but sounds like the only consequence is Java clients would have to explicitly set the idempotence key; from [the docs](https://kotlinlang.org/docs/reference/java-to-kotlin-interop.html#overloads-generation):

> a Kotlin function with default parameter values, it will be visible in Java only as a full signature, with all parameters present.

Since we can't use that annotation in interfaces, I'm okay with forcing Java apps to specify this argument (they already have to for the remaining args). If consistency is a concern, we can drop the default value and just force everyone to pass an explicit value (using docs to recommend a random UUID when no better candidate exists). Thoughts?

Reply:
the problem is existing kotlin clients will not have a JVM bytecode method of the signature `enqueue(QueueName, String, Duration?, Map)` anymore and fail to compile when the misk library is updated. if you add `@JvmOverloads`, it will generate the old JVM bytecode to be backwards compatbile.

Reply:
Can't do `@JvmOverloads` in an interface so I added another signature.